### ISSUE TITLE
Videos supported by avc hardware decoder(?) doesn't work in Android O

### ISSUE THREAD

**Issue by enriquezmark36** [remilia15/android_hardware_sprd#15]
Thanks for reading this, again. I've been trying to play a video with the mp4 format but it does not work. It works on nougat and older though. Also, recording camera videos or even opening the camera doesn't work :( .

I'm using the scx15 code instead of the sc8830 as the stock libraries _.sc8830.so_ naming suggests or else they won't compile.

---- 
Here's snippet of logcat (default + kernel) :
```
08-08 17:45:56.673  1778  1863 V SprdSimpleOMXComponent: Port 1 now enabled.
08-08 17:45:56.676  1778  1933 I SprdSimpleOMXComponent: fillThisBuffer, buffer: 0xb5423e60, header: 0xb5423b80, iRefCount: 0
08-08 17:45:56.676  1778  1933 I SPRDAVCDecoder: onQueueFilled, 937, outHeader:0xb5423b80, inHeader: 0xb54239a0, len: 31, nOffset: 0, time: 0, EOS: 0
08-08 17:45:56.676  1778  1933 V SPRDAVCDecoder: onQueueFilled, 977, dec_in.dataLen: 31, mPicId: 6
08-08 17:45:56.677  1778  1933 E MemoryHeapIon: Get_phy_addr_from_ion: Getphyaddr error!
08-08 17:45:56.677  1778  1933 V SPRDAVCDecoder: onQueueFilled, 1003, outHeader: 0xb5423b80, pBuffer: 0xb5419260, phyAddr: 0x0
08-08 17:45:56.677  1778  1933 D vndksupport: Loading /vendor/lib/hw/gralloc.sc8830.so from current namespace instead of sphal namespace.
08-08 17:45:56.679  1778  1933 V SPRDAVCDecoder: onQueueFilled, 1019, pBuffer: 0xb5419260, vaddr: 0xb20d5000
08-08 17:45:56.704     0     0 E         : [1:rd.h264.decoder: 1933] [c1] ion_phys: ion_phys is not implemented by this heap.
08-08 17:45:56.679  1778  1933 V SPRDAVCDecoder: onQueueFilled, 1022, yuv: b20d5000, mPicId: 6, outHeader: b5423b80, outHeader->pBuffer: b5419260, outHeader->nTimeStamp: 0
08-08 17:45:56.704     0     0 E         : [1:rd.h264.decoder: 1933] [c1] sprd_heap_ioctl alloc ret=0xffffffed error!
08-08 17:45:56.680  1778  1933 D VSP     : VSP_ACQUIRE_Dev, 199
08-08 17:45:56.680  1778  1933 D VSP     : H264DecDecode, 387, g_stream_offset: 4, g_slice_datalen: 27, g_nalu_ptr->len: 25, destuffing_num: 2
08-08 17:45:56.680  1778  1933 D VSP     : H264Dec_Read_SPS_PPS_SliceHeader, 74, the first byte: 0x67.
08-08 17:45:56.680  1778  1933 D VSP     : H264DecDecode_NALU, 908, finished decoding one NALU
08-08 17:45:56.680  1778  1933 D VSP     : H264DecDecode, 408, g_nalu_ptr->len: 25, frame_dec_finish: 0,ret:  -9 
08-08 17:45:56.680  1778  1933 D VSP     : H264DecDecode, 446, exit decoder, error_flag: 0
08-08 17:45:56.680  1778  1933 D VSP     : VSP_RELEASE_Dev, 257, ret: 0
08-08 17:45:56.680  1778  1933 I SPRDAVCDecoder: onQueueFilled, 1033, decRet: -9, 0ms, dec_out.frameEffective: 0, needIVOP: 1
08-08 17:45:56.680  1778  1933 D vndksupport: Loading /vendor/lib/hw/gralloc.sc8830.so from current namespace instead of sphal namespace.
08-08 17:45:56.682  1778  1933 I SPRDAVCDecoder: now, we don't take care of the decoder return: -9
08-08 17:45:56.683  1778  1933 I SPRDAVCDecoder: handlePortSettingChangeEvent, 1145, mWidth: 1920, mHeight: 1088, info->picWidth: 1920, info->picHeight: 1088
08-08 17:45:56.683  1778  1933 E SPRDAVCDecoder: set ddr freq to 500000khz
08-08 17:45:56.683  1778  1933 I SPRDAVCDecoder: handlePortSettingChangeEvent, 1154, info->numRefFrames: 3, info->has_b_frames: 1, def->nBufferCountActual: 6
08-08 17:45:56.683  1778  1933 I VSP     : H264Dec_ReleaseRefBuffers, 32
08-08 17:45:56.683  1778  1933 I SprdSimpleOMXComponent: fillThisBuffer, buffer: 0xb5423e68, header: 0xb5423bd0, iRefCount: 0
08-08 17:45:56.689  1778  1860 V SprdSimpleOMXComponent: Port 1 now disabled.
```
Then some lines after I also get this:
```
08-08 17:45:56.792     0     0 E         : [2:rd.h264.decoder: 1933] [c2] ion_phys: ion_phys is not implemented by this heap.
08-08 17:45:56.792     0     0 E         : [2:rd.h264.decoder: 1933] [c2] sprd_heap_ioctl alloc ret=0xffffffed error!
```

Then I thought of changing this line's ION_HEAP_TYPE_SYSTEM to ION_HEAP_TYPE_CUSTOM: https://github.com/remilia15/android_kernel_samsung_core33g/blob/lineage-15.1/arch/arm/mach-sc/devices-sc8830.c#L1154 . (_I'm not using this kernel source but they have similar content_)

The line above disappeared. Camera app can be opened, but still unable to record a video yet it can shoot pictures. Here's another short snippet of logcat (default + kernel):
```
08-08 17:27:29.267  1582  1838 V SPRDAVCDecoder: onQueueFilled, 1019, pBuffer: 0xa6619260, vaddr: 0xa322a000
08-08 17:27:29.267  1582  1838 V SPRDAVCDecoder: onQueueFilled, 1022, yuv: a322a000, mPicId: 21, outHeader: a6623b80, outHeader->pBuffer: a6619260, outHeader->nTimeStamp: 0
08-08 17:27:29.268  1582  1838 D VSP     : VSP_ACQUIRE_Dev, 199
08-08 17:27:29.268  1582  1838 D VSP     : H264DecDecode, 387, g_stream_offset: 4, g_slice_datalen: 27, g_nalu_ptr->len: 25, destuffing_num: 2
08-08 17:27:29.268  1582  1838 D VSP     : H264Dec_Read_SPS_PPS_SliceHeader, 74, the first byte: 0x67.
08-08 17:27:29.268  1582  1838 D VSP     : H264DecDecode_NALU, 908, finished decoding one NALU
08-08 17:27:29.268  1582  1838 D VSP     : H264DecDecode, 408, g_nalu_ptr->len: 25, frame_dec_finish: 0,ret:  -9 
08-08 17:27:29.268  1582  1838 D VSP     : H264DecDecode, 446, exit decoder, error_flag: 0
08-08 17:27:29.268  1582  1838 D VSP     : VSP_RELEASE_Dev, 257, ret: 0
08-08 17:27:29.268  1582  1838 I SPRDAVCDecoder: onQueueFilled, 1033, decRet: -9, 0ms, dec_out.frameEffective: 0, needIVOP: 1
08-08 17:27:29.268  1582  1838 D vndksupport: Loading /vendor/lib/hw/gralloc.sc8830.so from current namespace instead of sphal namespace.
08-08 17:27:29.269     0     0 W         : [0:   Binder:262_2:  279] [c0] ion: malloc from cma mem: size=002fd000,phy addr=9d700000, time=15ms
08-08 17:27:29.269  1582  1838 I SPRDAVCDecoder: now, we don't take care of the decoder return: -9
```

**Comment by remilia15** [comment]
Camera recording never work on O with prebuilt HAL. The reason was we need to switch to native handle implementation rather than getting buffers from gralloc HAL

**Comment by enriquezmark36** [comment]
How about playing a video, it seems that it also uses gralloc to fetch buffers, doesn't it?

**Comment by remilia15** [comment]
Playing a video is ok. 1080p30 is fine

**Comment by remilia15** [comment]
Other resolutions too

**Comment by enriquezmark36** [comment]
**TL;DR;** It's okay now. I can now play videos up to 1080p30. So, I'll be closing this. :)

It seems that a patch `libstagefright: Make it possible to skip OMX buffer reallocation`  didn't make it to LineageOS 15. Though this made me ask:
What does `BOARD_CANT_REALLOCATE_OMX_BUFFERS := true` in `scx35-common/BoardConfigCommon.mk` even do? 

Unfortunately, it doesn't fix the camera recording. It hangs after the 00:01 time. AFAIK, LineageOS devs are working on the "legacy" HAL1 camera recording issue.

Anyways, here's the I used patch on `frameworks/av` that just tests this idea.
```
diff --git a/media/libstagefright/ACodec.cpp b/media/libstagefright/ACodec.cpp
index def8ada3c..47d622bfa 100644
--- a/media/libstagefright/ACodec.cpp
+++ b/media/libstagefright/ACodec.cpp
@@ -1182,6 +1182,12 @@ status_t ACodec::configureOutputBuffersFromNativeWindow(
     // 2. try to allocate two (2) additional buffers to reduce starvation from
     //    the consumer
     //    plus an extra buffer to account for incorrect minUndequeuedBufs
+#if 1
+    // Some devices don't like to set OMX_IndexParamPortDefinition at this
+    // point (even with an unmodified def), so skip it if possible.
+    // This check was present in KitKat.
+    if (def.nBufferCountActual < def.nBufferCountMin + *minUndequeuedBuffers) {
+#endif
     for (OMX_U32 extraBuffers = 2 + 1; /* condition inside loop */; extraBuffers--) {
         OMX_U32 newBufferCount =
             def.nBufferCountMin + *minUndequeuedBuffers + extraBuffers;
@@ -1201,7 +1207,9 @@ status_t ACodec::configureOutputBuffersFromNativeWindow(
             return err;
         }
     }
-
+#if 1
+    }
+#endif
```

**Comment by remilia15** [comment]
The flag is removed in LOS 15.1
However, you can bring it back via patching method

**Comment by enriquezmark36** [comment]
Oh, that's why. Does the camera recording work by now for HAL1? I don't think I have done much researches.
